bundles/tools.vitruv.change.changederivation/src/tools/vitruv/change/changederivation/AbstractStateBasedChangeResolutionStrategy.xtend

```diff
@@ -0,0 +1,146 @@
+        newState.checkNoProxies("new state")
+        // It is possible that root elements are automatically generated during resource creation (e.g., Java packages).
+        // Thus, we create the resource and then monitor the re-insertion of the elements
+        val monitoredResourceSet = withGlobalFactories(new ResourceSetImpl());
+        val newResource = monitoredResourceSet.createResource(newState.URI)
+        newResource.contents.clear()
+        return newResource.record [
+            newResource.contents += EcoreUtil.copyAll(newState.contents)
+        ]
+    }
+
+    override getChangeSequenceForDeleted(Resource oldState) {
+        checkArgument(oldState !== null, "old state must not be null!")
+        oldState.checkNoProxies("old state")
+        // Setup resolver and copy state:
+        val monitoredResourceSet = new ResourceSetImpl()
+        val currentStateCopy = ResourceCopier.copyViewResource(oldState, monitoredResourceSet)
+        return currentStateCopy.record [
+            currentStateCopy.contents.clear()
+        ]
+    }
+
+    private def <T extends Notifier> record(Resource resource, ()=>void function) {
+        try (val changeRecorder = new ChangeRecorder(resource.resourceSet)) {
+            changeRecorder.beginRecording
+            changeRecorder.addToRecording(resource)
+            function.apply()
+            val recordedChanges = changeRecorder.endRecording
+            assignIds(recordedChanges, resource.resourceSet)
+            return recordedChanges.unresolve
+        }
+    }
+    
+    private def void assignIds(VitruviusChange recordedChange, ResourceSet resourceSet) {
+    	val changes = recordedChange.EChanges
+    	val idResolver = IdResolver.create(resourceSet)
+    	val eChangeIdManager = new EChangeIdManager(idResolver)
+    	changes.toList.reverseView.forEach[applyBackward]
+		changes.forEach[ change |
+			eChangeIdManager.setOrGenerateIds(change)
+			change.applyForward(idResolver)
+		]
+    }
+    
+    /**
+     * Compares states using the EMFCompareInstance provided by {@link #getEMFCompareInstance()} and replays the changes to the current state.
+     */
+    private def compareStatesAndReplayChanges(Notifier newState, Notifier currentState) {
+        val scope = new DefaultComparisonScope(newState, currentState, null)
+        val differences = buildEMFCompare().compare(scope).differences
+        // Replay the EMF compare differences
+        val mergerRegistry = IMerger.RegistryImpl.createStandaloneInstance()
+        val merger = new BatchMerger(mergerRegistry)
+        merger.copyAllLeftToRight(differences, new BasicMonitor)
+    }
+    
+    private def buildEMFCompare() {
+    	val usedMatchEngineFactoryRegistry = MatchEngineFactoryRegistryImpl.createStandaloneInstance()
+    	getMatchEngineFactories.forEach[it | it.ranking = 20]
+    	getMatchEngineFactories.forEach[it | usedMatchEngineFactoryRegistry.add(it)]
```

Comment:
This will not work out because the list is fetched twice, i.e., if every method call generates a new list with new MatchEngineFactory instances (as the current default implementation), the rank of the used instances is not set. Either store the list in a variable or combine both calls to one forEach method.